### DISCUSSION TITLE
Broadcasting tracker data through network

### DISCUSSION THREAD

**Discussion by kenavolic** [PlusToolkit/PlusLib#883] [Q&A]
Hi,

In the same spirit as #871, I had to test the broadcasting of tracker data through network with _OpenIGTLinkTracker_ device.

Here is setup:

- PC1 with a server built from https://github.com/PlusToolkit/PlusLib/pull/882 streaming accelerometer data

```xml
<PlusConfiguration version="2.1">
    <DataCollection StartupDelaySec="1.0">
        <DeviceSet Name="PlusServer: Generic sensor" Description="Broadcasting available sensors tracking data through OpenIGTLink" />
        <Device Id="TrackerDevice" Type="GenericSensor" ToolReferenceFrame="Tracker">
            <DataSources>
                <DataSource Type="Tool" Id="Accelerometer" PortName="Accelerometer" />
            </DataSources>
            <OutputChannels>
                <OutputChannel Id="TrackerStream">
                    <DataSource Id="Accelerometer"/>
                </OutputChannel>
            </OutputChannels>
        </Device>
        <Device Id="CaptureDevice" Type="VirtualCapture" BaseFilename="RecordingTest.igs.mha" EnableCapturingOnStart="FALSE">
            <InputChannels>
                <InputChannel Id="TrackerStream" />
            </InputChannels>
        </Device>
    </DataCollection>
    <PlusOpenIGTLinkServer MaxNumberOfIgtlMessagesToSend="1" MaxTimeSpentWithProcessingMs="50" ListeningPort="18944" OutputChannelId="TrackerStream">
        <DefaultClientInfo>
            <MessageTypes>
                <Message Type="TRANSFORM" />
            </MessageTypes>
            <TransformNames>
                <Transform Name="AccelerometerToTracker" />
            </TransformNames>
        </DefaultClientInfo>
    </PlusOpenIGTLinkServer>
</PlusConfiguration>

```

- PC2 with a server collecting data from a _OpenIGTLinkTracker_ device taken

```xml
 <PlusConfiguration version="2.4">
   <DataCollection StartupDelaySec="1.0">
     <DeviceSet Name="PlusServer: Generic sensor" Description="Broadcasting available sensors tracking data through OpenIGTLink" />
     <Device
       Id="TrackerDevice"
       Type="OpenIGTLinkTracker"
       MessageType="TRANSFORM"
       ToolReferenceFrame="Tracker"
       ServerAddress="172.19.11.210"
       ServerPort="18944"
       IgtlMessageCrcCheckEnabled="true"
       UseReceivedTimestamps="false"
       AcquisitionRate="20"
       ReconnectOnReceiveTimeout="false">
       <DataSources>
         <DataSource Type="Tool" Id="Accelerometer" />
       </DataSources>
       <OutputChannels>
         <OutputChannel Id="TrackerStream">
           <DataSource Id="Accelerometer" />
         </OutputChannel>
       </OutputChannels>
     </Device>
   </DataCollection>
   <PlusOpenIGTLinkServer MaxNumberOfIgtlMessagesToSend="1" MaxTimeSpentWithProcessingMs="50" ListeningPort="18944" SendValidTransformsOnly="true" OutputChannelId="TrackerStream">
     <DefaultClientInfo>
       <MessageTypes>
         <Message Type="TRANSFORM" />
       </MessageTypes>
       <TransformNames>
         <Transform Name="AccelerometerToTracker" />
       </TransformNames>
     </DefaultClientInfo>
   </PlusOpenIGTLinkServer>
 </PlusConfiguration>
```

When I run the server on PC2, I get the following errors:

```
|WARNING|005.301800| Incoming TRANSFORM message was received without a tool status. Defaulting to TOOL_OK.| in /home/user/.conan/data/PlusLib/2.9.0/latim/stable/build/88131246a5a853bdb23bd09d9d3d766a01f9b668/source_subfolder/src/PlusOpenIGTLink/vtkPlusIgtlMessageCommon.cxx(845)
|INFO|005.301911| ToolTimeStampedUpdate failed for tool: AccelerometerToTrack with timestamp: 5.301888| in /home/user/.conan/data/PlusLib/2.9.0/latim/stable/build/88131246a5a853bdb23bd09d9d3d766a01f9b668/source_subfolder/src/PlusDataCollection/OpenIGTLink/vtkPlusOpenIGTLinkTracker.cxx(362)
```

**IMPORTANT NOTE**: Collecting the data by mean of a direct client connection from PC2 works perfectly.

This looks like the same issue related in #367 .

_Attached_: Debug logs
[PlusLog.txt](https://github.com/PlusToolkit/PlusLib/files/7802909/PlusLog.txt)

**Comment by lassoan** [comment]
This error indicates that the message name is truncated (`AccelerometerToTrack` instead of `AccelerometerToTracker`):

> 010322_171334.099|ERROR|000.163510| TrackerDevice: Failed to update tool - unable to find tool: AccelerometerToTrack| in /home/user/.conan/data/PlusLib/2.9.0/latim/stable/build/88131246a5a853bdb23bd09d9d3d766a01f9b668/source_subfolder/src/PlusDataCollection/vtkPlusDevice.cxx(1632)

In earlier OpenIGTLink versions device name was limited to 20 characters. Current Plus clients should be able to use longer device names but for some reason OpenIGTLink v1 messages are used:

> 010322_171334.993|DEBUG|001.057899| Data sent by default: IGTL version: 1Message types: TRANSFORM. TDATARequested: FALSE. LastTDATASentTimeStamp: -1. TDATAResolution: 0. . Transforms: AccelerometerToTracker. Strings: (none). Images: (none)| in /home/user/.conan/data/PlusLib/2.9.0/latim/stable/build/88131246a5a853bdb23bd09d9d3d766a01f9b668/source_subfolder/src/PlusServer/vtkPlusOpenIGTLinkServer.cxx(192)

To verify that this is indeed the root cause of the problem, you could shorten the `Device` `ToolReferenceFrame` or `DataSource` `Id` names. Then we could improve the error message and fix the OpenIGTLink sender/receiver to make sure they can switch to OpenIGTLink protocol version that supports long device names.

**Reply by kenavolic** [reply]
Thanks for the answer. I can confirm that shortening the name solves the error. I can now receive the transform. There is still warning messages in my logs though:

`|WARNING|010.851490| Incoming TRANSFORM message was received without a tool status. Defaulting to TOOL_OK.| in /home/user/.conan/data/PlusLib/2.9.0/latim/stable/build/88131246a5a853bdb23bd09d9d3d766a01f9b668/source_subfolder/src/PlusOpenIGTLink/vtkPlusIgtlMessageCommon.cxx(845)`
